src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs

```diff
@@ -65,14 +67,23 @@ public async virtual Task<HostSecretsInfo> GetHostSecretsAsync()
         {
             if (_hostSecrets == null)
             {
-                HostSecrets hostSecrets = await LoadSecretsAsync<HostSecrets>();
+                HostSecrets hostSecrets;
+                await _semaphoreSlim.WaitAsync();
+                try
+                {
+                    hostSecrets = await LoadSecretsAsync<HostSecrets>();
 
-                if (hostSecrets == null)
+                    if (hostSecrets == null)
+                    {
+                        // host secrets do not yet exist so generate them
+                        _logger.LogDebug(Resources.TraceHostSecretGeneration);
+                        hostSecrets = GenerateHostSecrets();
+                        await PersistSecretsAsync(hostSecrets);
+                    }
+                }
+                finally
                 {
-                    // host secrets do not yet exist so generate them
-                    _logger.LogDebug(Resources.TraceHostSecretGeneration);
-                    hostSecrets = GenerateHostSecrets();
-                    await PersistSecretsAsync(hostSecrets);
+                    _semaphoreSlim.Release();
```

Comment:
A few blocks down we check for stale keys and go through refresh/persist. Shouldn't this be within the lock as well to prevent concurrent writes?

Reply:
Lock is extended to whole method.